File: server/services/urlAnalyzer.js
import net from "node:net";

export function validateUrl(value) {
  let parsed;
  try { parsed = new URL(value); } catch { throw new Error("Enter a complete http:// or https:// website URL."); }
  if (!['http:', 'https:'].includes(parsed.protocol)) throw new Error("Only HTTP and HTTPS URLs can be analyzed.");
  const host = parsed.hostname.toLowerCase();
  if (host === 'localhost' || host.endsWith('.local') || isPrivateAddress(host)) throw new Error("Local or private network addresses cannot be analyzed.");
  return parsed;
}

function isPrivateAddress(host) {
  if (!net.isIP(host)) return false;
  return host === '::1' || host.startsWith('127.') || host.startsWith('10.') || host.startsWith('192.168.') || /^172\.(1[6-9]|2\d|3[01])\./.test(host);
}

export function analyzeUrl(value) {
  const parsed = validateUrl(value);
  const host = parsed.hostname.toLowerCase().replace(/\.$/, '');
  const labels = host.split('.');
  const isIpAddress = Boolean(net.isIP(host));
  const domain = isIpAddress ? host : labels.slice(-2).join('.');
  const specialCharacterCount = (value.match(/[^a-zA-Z0-9./?=&_%:-]/g) || []).length;
  const features = {
    urlLength: value.length, domainLength: domain.length,
    digitCount: (host.match(/\d/g) || []).length, hyphenCount: (host.match(/-/g) || []).length,
    dotCount: (host.match(/\./g) || []).length, subdomainCount: Math.max(0, labels.length - 2),
    specialCharacterCount, hasHttps: parsed.protocol === 'https:', isIpAddress,
    hasPunycode: host.includes('xn--'), hasUnicode: /[^\u0000-\u007f]/.test(host),
    suspiciousStructure: value.length > 100 || labels.length > 4 || (host.match(/-/g) || []).length >= 3 || value.includes('@')
  };
  return { parsed, domain, tld: isIpAddress ? null : labels.at(-1), features };
}

File: backend/analyzers/url_analyzer.py
"""Safe URL parsing and lexical feature extraction."""
from ipaddress import ip_address
from urllib.parse import urlparse
import re

def validate_public_url(url: str) -> None:
    parsed = urlparse(url)
    if parsed.scheme not in {"http", "https"} or not parsed.hostname:
        raise ValueError("Enter a complete http:// or https:// website URL.")
    host = parsed.hostname.lower().rstrip(".")
    if host in {"localhost"} or host.endswith(".local"):
        raise ValueError("Local network addresses cannot be analyzed.")
    try:
        address = ip_address(host)
        if address.is_private or address.is_loopback or address.is_link_local or address.is_reserved:
            raise ValueError("Private network addresses cannot be analyzed.")
    except ValueError as error:
        if "cannot be analyzed" in str(error):
            raise

def extract_url_features(url: str) -> dict:
    validate_public_url(url)
    parsed = urlparse(url)
    host = parsed.hostname.lower().rstrip(".")
    labels = host.split(".")
    try:
        ip_address(host)
        is_ip = True
    except ValueError:
        is_ip = False
    domain = host if is_ip else ".".join(labels[-2:]) if len(labels) >= 2 else host
    specials = len(re.findall(r"[^A-Za-z0-9./?=&_%:-]", url))
    return {
        "domain": domain, "hostname": host, "tld": "" if is_ip else labels[-1],
        "url_length": len(url), "domain_length": len(domain), "path_length": len(parsed.path),
        "query_length": len(parsed.query), "digit_count": sum(c.isdigit() for c in host),
        "hyphen_count": host.count("-"), "dot_count": host.count("."),
        "special_character_count": specials, "subdomain_count": max(0, len(labels) - 2),
        "has_https": int(parsed.scheme == "https"), "has_punycode": int("xn--" in host),
        "has_unicode_lookalike": int(any(ord(c) > 127 for c in host)), "is_ip_address": int(is_ip),
        "excessive_subdomains": int(len(labels) > 4),
        "suspicious_structure": int(len(url) > 100 or host.count("-") >= 3 or "@" in url),
    }
